drivers/tee/optee/call.c

```diff
@@ -28,28 +28,363 @@ static struct optee_session *find_session(struct optee_context_data *ctxdata,
+			goto exit_set_ret;
+
+		process_ocall_memrefs(msg_param, msg_num_params, true);
+		call_ctx->rpc_must_release = true;
+
+		cmd_id = (u32)call_ctx->rpc_arg->params[0].u.value.a;
+		ocall->u.value.a =
+			TEE_IOCTL_OCALL_MAKE_PAIR(TEE_IOCTL_OCALL_CMD_INVOKE,
+						  cmd_id);
+
+		clnt_id = &call_ctx->rpc_arg->params[1].u.value;
+		memcpy(&ocall->u.value.b, clnt_id, TEE_IOCTL_UUID_LEN);
+		break;
+	default:
+		/* NOT REACHED */
+		rc = -EINVAL;
+		goto exit_set_ret;
+	}
+
+	return rc;
+
+exit_set_ret:
+	call_ctx->rpc_arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+	call_ctx->rpc_arg->ret_origin = TEEC_ORIGIN_COMMS;
+	return rc;
+}
+
+/* Requires the corresponding binary session semaphore to be held */
+static int process_ocall_reply(u32 ret, u32 ret_origin,
+			       struct tee_param *params, u32 num_params,
+			       struct tee_param *ocall,
+			       struct optee_call_ctx *call_ctx)
+{
+	const u64 func = tee_param_get_ocall_func(ocall);
+	struct tee_shm *shm;
+	void *shm_pages_list;
+	struct optee_msg_param *msg_param;
+	u32 msg_num_params;
+	int rc = 0;
+
+	rc = verify_ocall_reply(func, params, num_params, call_ctx);
+	if (rc)
+		goto exit_set_ret;
+
+	switch (func) {
+	case TEE_IOCTL_OCALL_CMD_SHM_ALLOC:
+		if (ret != TEEC_SUCCESS)
+			goto exit_propagate_ret;
+
+		shm = tee_shm_get_from_id(call_ctx->ctx,
+					  (int)params[0].u.value.c);
+		if (IS_ERR(shm)) {
+			rc = PTR_ERR(shm);
+			goto exit_set_ret;
+		}
+
+		rc = optee_rpc_process_shm_alloc(shm, call_ctx->rpc_arg->params,
+						 &shm_pages_list);
+
+		/* The CA holds a reference */
+		tee_shm_put(shm);
+
+		if (rc)
+			goto exit_set_ret;
+
+		if (shm_pages_list)
+			optee_free_pages_list(shm_pages_list, shm->num_pages);
+		break;
+	case TEE_IOCTL_OCALL_CMD_SHM_FREE:
+		/* OP-TEE ignores ret and ret_origin for an SHM_FREE RPC */
+		if (ret != TEEC_SUCCESS)
+			goto exit_propagate_ret;
+		/* NOTHING */
+		break;
+	case TEE_IOCTL_OCALL_CMD_INVOKE:
+		/* -2 here and +2 below to skip the OCALL descriptors */
+		msg_num_params = call_ctx->rpc_arg->num_params - 2;
+		if (num_params < msg_num_params) {
+			rc = -EINVAL;
+			goto exit_set_ret;
+		}
+
+		msg_param = call_ctx->rpc_arg->params + 2;
+		rc = optee_to_msg_param(msg_param, msg_num_params, params);
+		if (rc)
+			goto exit_set_ret;
+
+		process_ocall_memrefs(msg_param, msg_num_params, false);
+		call_ctx->rpc_must_release = false;
+
+		call_ctx->rpc_arg->params[0].u.value.b = ret;
+		call_ctx->rpc_arg->params[0].u.value.c = ret_origin;
+		break;
+	default:
+		rc = -EINVAL;
+		goto exit_set_ret;
+	}
+
+	call_ctx->rpc_arg->ret = TEEC_SUCCESS;
+	call_ctx->rpc_arg->ret_origin = TEEC_ORIGIN_COMMS;
+
+	return rc;
+
+exit_propagate_ret:
+	call_ctx->rpc_arg->ret = ret;
+	call_ctx->rpc_arg->ret_origin = ret_origin;
+	return -EINVAL;
+exit_set_ret:
+	call_ctx->rpc_arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+	call_ctx->rpc_arg->ret_origin = TEEC_ORIGIN_COMMS;
+	return rc;
+}
+
+static inline void clear_call_ctx(struct optee_call_ctx *call_ctx)
```

Comment:
No need for `inline`, leave it to the compiler.

Reply:
Removed.